sdk/identity/identity/src/credentials/managedIdentityCredential/utils.ts

```diff
@@ -72,3 +73,33 @@ export function parseExpiresOn(body: TokenResponseParsedBody): number {
     `Failed to parse token expiration from body. expires_in="${body.expires_in}", expires_on="${body.expires_on}"`
   );
 }
+
+/**
+ * Given a token response, return the timestamp for refreshing token as the number of milliseconds from the Unix epoch.
+ * @param body - A parsed response body from the authentication endpoint.
+ */
+export function parseRefreshTimestamp(body: TokenResponseParsedBody): number {
+  if (typeof body.refresh_in === "number") {
+    return Date.now() + body.refresh_in * 1000;
+  } else {
+    const durationInMilliseconds = parseExpirationTimestamp(body) - Date.now();
+    const duration = formatTime(durationInMilliseconds);
+    if (duration.hours >= 2) {
+      return Date.now() + durationInMilliseconds / 2;
+    } else {
+      return Date.now() + durationInMilliseconds;
+    }
+  }
+}
+
+function formatTime(milliseconds: number) {
+  const seconds = Math.floor((milliseconds / 1000) % 60);
+  const minutes = Math.floor((milliseconds / 1000 / 60) % 60);
+  const hours = Math.floor((milliseconds / 1000 / 60 / 60) % 24);
```

Comment:
Feels a little sketchy to wrap hours modulo 24. Is there any risk of tokens that could live for >24h that are then treated as if they should be refreshed very soon?

Reply:
You do raise a good point here. Though the MSI tokens usually last for 24 hours.

Reply:
I changed the logic to parse the durationInHours as an inline computation to just compute from milliseconds / 3600/1000. We anyway don't need the other fields for now.